File: utils/alert_module.py
"""告警模块分类 —— 加载 config/alert_modules.yaml 并暴露判定函数。

跨端单一真实源：
  - 本模块：Python 运行时直接加载 yaml。
  - 前端：frontend/src/lib/alert-module.config.generated.ts
         由 frontend/scripts/gen-alert-module-config.mts 读同一份 yaml 生成。
"""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any

import yaml

_CONFIG_PATH = Path(__file__).resolve().parent.parent / "config" / "alert_modules.yaml"


@lru_cache(maxsize=1)
def _load() -> dict[str, Any]:
    with _CONFIG_PATH.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    if not isinstance(data, dict):
        raise ValueError(f"Invalid alert_modules.yaml: {type(data)}")
    return data


def get_modules_in_order() -> list[str]:
    return list(_load().get("modules_in_order", []))


def get_grain_map() -> dict[str, str]:
    return dict(_load().get("grain_map", {}))


def get_grain_options() -> list[str]:
    return list(_load().get("grain_options", []))


def grain_label(target_level: str | None) -> str:
    """target_level → 中文粒度标签；未知值原样返回。"""
    text = (target_level or "").strip()
    return get_grain_map().get(text, text)


def infer_alert_module(target_key: str | None, rule_code: str | None) -> str:
    """从告警的 target_key / rule_code 推断所属模块。

    规则配置：config/alert_modules.yaml，按 rules 数组顺序匹配，命中即返回。
    """
    tk_raw = target_key or ""
    rc_raw = rule_code or ""
    tk_upper = tk_raw.upper()
    rc_upper = rc_raw.upper()

    cfg = _load()
    for rule in cfg.get("rules", []):
        match = rule.get("match") or {}
        # rule_code_upper_includes
        for sub in match.get("rule_code_upper_includes", []) or []:
            if sub and sub.upper() in rc_upper:
                return rule["module"]
        # target_key_upper_startswith
        for prefix in match.get("target_key_upper_startswith", []) or []:
            if prefix and tk_upper.startswith(prefix.upper()):
                return rule["module"]
        # target_key_upper_includes
        for sub in match.get("target_key_upper_includes", []) or []:
            if sub and sub.upper() in tk_upper:
                return rule["module"]
        # target_key_contains (大小写敏感原串)
        for sub in match.get("target_key_contains", []) or []:
            if sub and sub in tk_raw:
                return rule["module"]
        # target_key_equals
        for val in match.get("target_key_equals", []) or []:
            if tk_raw == val:
                return rule["module"]

    return cfg.get("default_module", "其他")

File: frontend/src/lib/alert-module.ts
/**
 * 告警模块 / 粒度推断与中文化。
 *
 * 单一真实源：`config/alert_modules.yaml`。
 *   - Python 运行时加载（utils/alert_module.py）。
 *   - TS 由 scripts/gen_alert_module_config.py 生成 `alert-module.config.generated.ts`。
 *
 * 改规则只改 yaml；改完跑 `pnpm gen:config`。
 */

import { ALERT_MODULE_CONFIG } from "./alert-module.config.generated";
import type { AlertModuleRule } from "./alert-module.types";

export type AlertModule =
  | "外检"
  | "内检"
  | "新人外检"
  | "新人内检"
  | "BadCase"
  | "全局"
  | "其他";

export const ALERT_MODULES_IN_ORDER: AlertModule[] =
  ALERT_MODULE_CONFIG.modules_in_order.slice() as AlertModule[];

export const ALERT_GRAIN_OPTIONS = ALERT_MODULE_CONFIG.grain_options;
export type AlertGrainLabel = (typeof ALERT_GRAIN_OPTIONS)[number];

const DEFAULT_MODULE = ALERT_MODULE_CONFIG.default_module as AlertModule;

function matchRule(rule: AlertModuleRule, tkRaw: string, rcRaw: string): boolean {
  const tkUpper = tkRaw.toUpperCase();
  const rcUpper = rcRaw.toUpperCase();
  const m = rule.match;

  if (m.rule_code_upper_includes) {
    for (const sub of m.rule_code_upper_includes) {
      if (sub && rcUpper.includes(sub.toUpperCase())) return true;
    }
  }
  if (m.target_key_upper_startswith) {
    for (const prefix of m.target_key_upper_startswith) {
      if (prefix && tkUpper.startsWith(prefix.toUpperCase())) return true;
    }
  }
  if (m.target_key_upper_includes) {
    for (const sub of m.target_key_upper_includes) {
      if (sub && tkUpper.includes(sub.toUpperCase())) return true;
    }
  }
  if (m.target_key_contains) {
    for (const sub of m.target_key_contains) {
      if (sub && tkRaw.includes(sub)) return true;
    }
  }
  if (m.target_key_equals) {
    for (const val of m.target_key_equals) {
      if (tkRaw === val) return true;
    }
  }
  return false;
}

/** 从 target_key / rule_code 推断所属业务模块。规则源：config/alert_modules.yaml。 */
export function inferAlertModule(
  targetKey: string | null | undefined,
  ruleCode: string | null | undefined,
): AlertModule {
  const tk = (targetKey ?? "").toString();
  const rc = (ruleCode ?? "").toString();

  for (const rule of ALERT_MODULE_CONFIG.rules) {
    if (matchRule(rule, tk, rc)) {
      return rule.module as AlertModule;
    }
  }
  return DEFAULT_MODULE;
}

/** `target_level` → 中文粒度标签；未知值原样返回。 */
export function grainLabel(targetLevel: string | null | undefined): string {
  const text = (targetLevel ?? "").toString();
  return ALERT_MODULE_CONFIG.grain_map[text] ?? text;
}
